features/netsocket/nsapi_dns.cpp

```diff
@@ -194,6 +273,109 @@ static int dns_scan_response(const uint8_t **p, nsapi_addr_t *addr, unsigned add
     return count;
 }
 
+static void nsapi_dns_cache_add(const char *host, SocketAddress *address, uint32_t ttl)
+{
+    // RFC 1034: if TTL is zero, entry is not added to cache
+    if (!ttl) {
+        return;
+    }
+
+    // Checks if already cached
+    if (nsapi_dns_cache_find(host, address->get_ip_version(), NULL) == NSAPI_ERROR_OK) {
+        return;
+    }
+
+    dns_cache_mutex.lock();
+
+    int index = -1;
+    uint64_t accessed = -1;
+
+    // Finds free or last accessed entry
+    for (int i = 0; i < DNS_CACHE_SIZE; i++) {
+        if (!dns_cache[i]) {
+            index = i;
+            break;
+        } else if (dns_cache[i]->accessed <= accessed) {
+            accessed = dns_cache[i]->accessed;
+            index = i;
+        }
+    }
+
+    if (index < 0) {
+        return;
+    }
+
+    // Allocates in case entry is free, otherwise reuses
+    if (!dns_cache[index]) {
+        dns_cache[index] = new DNS_CACHE;
```

Comment:
This will explode on allocation failure - probably want `new (nothrow) DNS_CACHE` here with a null check.